### PR TITLE
ibmcloud delete-service-id command

### PR DESCRIPTION
**Pull request by mkumatag**
Implementation of `ccoctl ibmcloud delete-service-id`

### CODE REVIEW

File: pkg/cmd/provisioning/ibmcloud/create_service_id.go [hunk 1]
```diff
@@ -54,7 +54,7 @@ func NewCreateServiceIDCmd() *cobra.Command {
 
 	createServiceIDCmd.PersistentFlags().StringVar(&CreateOpts.Name, "name", "", "User-defined name for all created IBM Cloud resources (can be separate from the cluster's infra-id)")
 	createServiceIDCmd.MarkPersistentFlagRequired("name")
-	createServiceIDCmd.PersistentFlags().StringVar(&CreateOpts.CredRequestDir, "credentials-requests-dir", "", "Directory containing files of CredentialsRequests to create IAM Roles for (can be created by running 'oc adm release extract --credentials-requests --cloud=aws' against an OpenShift release image)")
+	createServiceIDCmd.PersistentFlags().StringVar(&CreateOpts.CredRequestDir, "credentials-requests-dir", "", "Directory containing files of CredentialsRequests to create IAM Roles for (can be created by running 'oc adm release extract --credentials-requests --cloud=ibmcloud' against an OpenShift release image)")
```
Comment (joelddiaz) [inline pkg/cmd/provisioning/ibmcloud/create_service_id.go:57]: This is definitely an improvement, but just out of curiosity: is this already implemented in the recent builds of the oc CLI?

File: pkg/cmd/provisioning/ibmcloud/delete_service_id.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
+package ibmcloud
+
+import (
+	"fmt"
+	"github.com/openshift/cloud-credential-operator/pkg/ibmcloud"
+	"github.com/pkg/errors"
+
+	//"github.com/pkg/errors"
+	"github.com/spf13/cobra"
```
Comment (joelddiaz) [inline pkg/cmd/provisioning/ibmcloud/delete_service_id.go:9]:
```suggestion
	"fmt"

	"github.com/pkg/errors"
	"github.com/spf13/cobra"
	
	"github.com/openshift/cloud-credential-operator/pkg/ibmcloud"
```
Just some organization around the import list.

File: pkg/cmd/provisioning/ibmcloud/delete_service_id.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	deleteServiceIDCmd := &cobra.Command{
+		Use:              "delete-service-id",
+		Short:            "Delete Service ID",
+		RunE:             deleteServiceIDCmd,
+		PersistentPreRun: initEnvForCreateServiceIDCmd,
+	}
+
+	deleteServiceIDCmd.PersistentFlags().StringVar(&CreateOpts.Name, "name", "", "User-defined name for all created IBM Cloud resources (can be separate from the cluster's infra-id)")
+	deleteServiceIDCmd.MarkPersistentFlagRequired("name")
+	deleteServiceIDCmd.PersistentFlags().StringVar(&CreateOpts.CredRequestDir, "credentials-requests-dir", "", "Directory containing files of CredentialsRequests to delete IAM Roles for (can be created by running 'oc adm release extract --credentials-requests --cloud=ibmcloud' against an OpenShift release image)")
```
Comment (joelddiaz) [inline pkg/cmd/provisioning/ibmcloud/delete_service_id.go:23]: Up until now, our `delete` subcommands have not required providing the list of CredentialsRequests. I'm not fundamentally opposed to this, but I'd like to hear from the rest of the team (cc @akhil-rane ).

File: pkg/cmd/provisioning/ibmcloud/service_id.go [hunk 4]
```diff
@@ -247,6 +250,67 @@ func (s *ServiceID) UnDo(targetDir string) error {
...
+
+	if len(allrecs) > 1 && !force {
+		return fmt.Errorf("more than one ServiceIDs present with %s name, please run with --force flag to delete all the entries forcefully", s.name)
+	} else if len(allrecs) == 0 {
+		log.Printf("no ServiceID found with name: %s", s.name)
+		return nil
+	} else {
+		if force {
+			log.Printf("--force flag present, will delete all the entries with %s name forcefully", s.name)
+		}
```
Comment (joelddiaz) [inline pkg/cmd/provisioning/ibmcloud/service_id.go:287]: Under this else clause, I don't see any reference to the `force` flag, so is this message really necessary?

Comment (mkumatag) [reply pkg/cmd/provisioning/ibmcloud/service_id.go:287]: This is just an FYI message to the user.,

File: pkg/cmd/provisioning/ibmcloud/create_service_id.go (hunk 1 above)
Comment (mkumatag) [reply pkg/cmd/provisioning/ibmcloud/create_service_id.go:57]: yes, it is already in via - https://github.com/openshift/oc/commit/482829c1362784b6e4ea9a6ba3f6f04ac8235a8e

File: pkg/cmd/provisioning/ibmcloud/delete_service_id.go (hunk 3 above)
Comment (akhil-rane) [reply pkg/cmd/provisioning/ibmcloud/delete_service_id.go:23]: As we cannot tag the service ids, we need to construct fullname to delete the service ids. This looks like a better approach to me.

**Review by akhil-rane — COMMENTED** [review]
These PRs (ibm ccoctl) do not have a corresponding jira ticket. I think it will be better to have one so that it can also go through qe testing.

File: pkg/cmd/provisioning/ibmcloud/delete_service_id.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+// NewDeleteServiceIDCmd provides the "delete-service-id" subcommand
+func NewDeleteServiceIDCmd() *cobra.Command {
+	deleteServiceIDCmd := &cobra.Command{
+		Use:              "delete-service-id",
+		Short:            "Delete Service ID",
+		RunE:             deleteServiceIDCmd,
+		PersistentPreRun: initEnvForCreateServiceIDCmd,
+	}
+
+	deleteServiceIDCmd.PersistentFlags().StringVar(&CreateOpts.Name, "name", "", "User-defined name for all created IBM Cloud resources (can be separate from the cluster's infra-id)")
```
[the remark this answers is not in the document]
Comment (mkumatag) [reply pkg/cmd/provisioning/ibmcloud/delete_service_id.go:21]: both these commands share few options, wondering if we can make this struct generic like `options` or so.!

File: pkg/cmd/provisioning/ibmcloud/delete_service_id.go (hunk 2 above)
Comment (mkumatag) [reply pkg/cmd/provisioning/ibmcloud/delete_service_id.go:9]: this is a miss from my end, will fix this.

File: pkg/cmd/provisioning/ibmcloud/delete_service_id.go [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+	"github.com/spf13/cobra"
+)
+
+// NewDeleteServiceIDCmd provides the "delete-service-id" subcommand
+func NewDeleteServiceIDCmd() *cobra.Command {
+	deleteServiceIDCmd := &cobra.Command{
+		Use:              "delete-service-id",
+		Short:            "Delete Service ID",
+		RunE:             deleteServiceIDCmd,
+		PersistentPreRun: initEnvForCreateServiceIDCmd,
```
[the remark this answers is not in the document]
Comment (mkumatag) [reply pkg/cmd/provisioning/ibmcloud/delete_service_id.go:18]: this is not needed, I will remove it from here, good catch.

**Comment by mkumatag** [comment]
> These PRs (ibm ccoctl) do not have a corresponding jira ticket. I think it will be better to have one so that it can also go through qe testing.

Not sure if I have access to your Jira, will create one if needed.

File: pkg/cmd/provisioning/ibmcloud/delete_service_id.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (akhil-rane) [reply pkg/cmd/provisioning/ibmcloud/delete_service_id.go:21]: yes, common sounds better. We can split it out later if things change.

**Comment by akhil-rane** [comment]
> > These PRs (ibm ccoctl) do not have a corresponding jira ticket. I think it will be better to have one so that it can also go through qe testing.
> 
> Not sure if I have access to your Jira, will create one if needed.

Our jira is public. I have created the jira issues for these tasks. You can edit and put additional info in the description.

https://issues.redhat.com/browse/CCO-143
https://issues.redhat.com/browse/CCO-144

**Comment by mkumatag** [comment]
@joelddiaz @akhil-rane addressed the review comments, ptal recent commit

**Review by joelddiaz — COMMENTED** [review]
:+1: will let @akhil-rane check that the items he brought up during review are addressesd, and then I think we're good to merge.

**Comment by akhil-rane** [comment]
/lgtm

@joelddiaz needs approve label

**Comment by joelddiaz** [comment]
we need a whole bunch of labels now :)
/approve
/assign @lwan-wanglin @jeana-redhat @sferich888

**Comment by lwan-wanglin** [comment]
@mkumatag I drafted the test cases for the pr https://url.corp.redhat.com/3ac9e4f, could you help please review it? thanks!
And i have a small question, would you mind telling me which situation will need to use --force parameter when run delete-service-id command?

**Comment by mkumatag** [comment]
> @mkumatag I drafted the test cases for the pr https://url.corp.redhat.com/3ac9e4f, could you help please review it? thanks!
> And i have a small question, would you mind telling me which situation will need to use --force parameter when run delete-service-id command?

I'm not able to login to that mentioned site with my RH ID, not sure if I need to register somewhere for that(accessing for the first time I feel). `--force` will be helpful when there are multiple service-ids present with the same name(this happens when someone creates the service-ids with the same name in the backend(ibmcloud) manually), for [more info](https://cloud.ibm.com/docs/account?topic=account-serviceids&interface=ui)

**Comment by lwan-wanglin** [comment]
> I'm not able to login to that mentioned site with my RH ID, not sure if I need to register somewhere for that(accessing for the first time I feel).
I know hive team has a share account, @joelddiaz could you share it to @mkumatag 

> `--force` will be helpful when there are multiple service-ids present with the same name(this happens when someone creates the service-ids with the same name in the backend(ibmcloud) manually), for [more info](https://cloud.ibm.com/docs/account?topic=account-serviceids&interface=ui)
thanks for your reply

**Comment by joelddiaz** [comment]
@lwan-wanglin I'm not aware of a hive-specific login. I've been using the same internal login that was shared with me a loooooong time ago. And based on the username/password, it does not appear to be specific to hive. Is there a process to get accounts into this test-case site so that we can hopefully use properly provisioned users when accessing/reviewing the information?

**Comment by lwan-wanglin** [comment]
@akhil-rane @mkumatag  I comfirmed with my manager, there is a share dev account , I will send it to you by email.

**Comment by mkumatag** [comment]
> @akhil-rane @mkumatag I comfirmed with my manager, there is a share dev account , I will send it to you by email.

I got the access details via mail and test cases lgtm.

**Comment by mkumatag** [comment]
Seems like we need one more label `px-approved` @joelddiaz @akhil-rane any idea who can help us getting this label?

**Comment by joelddiaz** [comment]
> Seems like we need one more label `px-approved` @joelddiaz @akhil-rane any idea who can help us getting this label?

That would be @sferich888 for the px approval

**Comment by mkumatag** [comment]
> > Seems like we need one more label `px-approved` @joelddiaz @akhil-rane any idea who can help us getting this label?
> 
> That would be @sferich888 for the px approval

@sferich888 can you please help getting `px-approved` label to this PR?

**Comment by mkumatag** [comment]
UT is missing, will submit a different PR for the same.

**Pull request merged**